File: server/src/db/entities/Job.ts
import { Entity, Column, ManyToOne, OneToMany } from "typeorm";

import BaseEntity from "./BaseEntity";
import Account from "./Account";
import Customer from "./Customer";
import Employee from "./Employee";
import JobDetail from "./JobDetail";
import JobHistory from "./JobHistory";

type JobType = "STINGING_BASIC" | "STRINGING_HYBRID";

type JobStatus = "PENDING" | "FINISHED";

@Entity()
export default class Job extends BaseEntity {
  @Column({ type: "text", nullable: false, default: "PENDING" })
  status: JobStatus;

  @Column({ type: "text", nullable: false })
  type: JobType;

  @Column({ type: "timestamptz", nullable: true })
  completedAt?: Date;

  // Relationships

  @Column()
  accountId: string;
  @ManyToOne((type) => Account, (c) => c.jobs)
  account: Account;

  @Column()
  customerId: string;
  @ManyToOne((type) => Customer, (c) => c.jobs)
  customer: Customer;

  @Column({ nullable: true })
  completedByEmployeeId?: string;
  @ManyToOne((type) => Employee, (e) => e.jobsCompleted, { nullable: true })
  completedByEmployee?: Employee;

  @OneToMany((type) => JobHistory, (j) => j.job)
  history: JobHistory[];

  @OneToMany((type) => JobDetail, (j) => j.job)
  details: JobDetail[];
}

File: Models/Job.cs
using System;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace TennisShopGuru.Models
{
  public enum JobType
  {
    STRINGING_BASIC,
    STRINGING_HYBRID
  }

  public enum JobStatus
  {
    PENDING,
    FINISHED
  }

  public class Job : BaseEntity
  {
    public int Id { get; set; }
    public JobStatus Status { get; set; }
    public JobType Type { get; set; }
    public DateTime CompletedAt { get; set; }

    // Relationships

    public int CompanyID { get; set; }
    [ForeignKey("CompanyID")]
    public Company Company { get; set; }

    public int CustomerID { get; set; }
    [ForeignKey("CustomerID")]
    public Customer Customer { get; set; }

    public int UserID { get; set; }
    [ForeignKey("UserID")]
    public User User { get; set; }
  }
}
